### PR TITLE
FPE Changes

### PR DESCRIPTION
**Pull request by HariharanIT**
# Related Issue
- Mention your issue description here
- clickup ticket - 
   https://app.clickup.com/t/8556478/APP-901 
   https://app.clickup.com/t/8556478/APP-902

# Propossed changes/Fix
- Created fpe-api folder and files which used to expose the method to install the components and reusing existing cmp/subcmp
-Created test-fpe used to check the installapi method
-FPEapi - in-progress

# Additional Info 
- Any additional info or context for the reviewer

# Checklist
- [ ] Tested on local/dev branch for all major platforms (Android, IOS, Desktop, Web).
- [ ] No commented out code
- [ ] Is any third party library, service used
- [ ] Tests
- [ ] If this change requires updates outside of the code, like updates in core, react-ui-kit, RTM/RTC configure.

# Dependent PRs 
- Mention dependent PR links.

# Screenshots

Original                |          Updated
:---------------------:  | :-----------------------:
**original screenshot** |  **updated screenshot**

### CODE REVIEW

**Comment by technophilic** [comment]
`test-fpe` should not be committed inside the core repository.

**Review by technophilic — CHANGES_REQUESTED** [review]

File: template/src/App.tsx [hunk 1]
```diff
@@ -39,8 +35,13 @@ if (Platform.OS === 'ios') {
 }
 
 const App: React.FC = () => {
-  const [phrase, onChangePhrase] = useState('');
-
+  const [appRoutes, setAppRoutes] = useState<CustomRoutesInterface[]>()
+  const [pluginReady, setPluginReady] = useState(false)
+  useEffect(( ) => {
+    installPlugin()
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/App.tsx]: Now install method move to seperate file

File: template/src/pages/VideoCall/index.tsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  recordingActive: boolean,
+  sidePanel: SidePanelType,
+  setSidePanel: React.Dispatch<SetStateAction<SidePanelType>>,
+  layout: any,
+  setLayout: React.Dispatch<SetStateAction<any>>,
+  isHost: boolean,
+  title: string,
+}
+const VideoCallContext = React.createContext((null as unknown) as VideoCallContextType);
+const VideoCallProvider = (props:any) => {
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/pages/VideoCall/index.tsx]: done

File: template/src/pages/VideoCall/index.tsx [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  setRecordingActive: any,
+  recordingActive: boolean,
+  sidePanel: SidePanelType,
+  setSidePanel: React.Dispatch<SetStateAction<SidePanelType>>,
+  layout: any,
+  setLayout: React.Dispatch<SetStateAction<any>>,
+  isHost: boolean,
+  title: string,
+}
+const VideoCallContext = React.createContext((null as unknown) as VideoCallContextType);
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/pages/VideoCall/index.tsx]: done

File: template/src/pages/Authenticate.tsx [hunk 4]
```diff
@@ -9,27 +9,25 @@
...
 
 const Authenticate = () => {
-  const [dim, setDim] = useState([
-    Dimensions.get('window').width,
-    Dimensions.get('window').height,
-    Dimensions.get('window').width > Dimensions.get('window').height,
-  ]);
+  const {getDimensionData}= useContext(DimensionContext)
+  const [isDesktop, setIsDesktop] = useState(false)
   let onLayout = (e: any) => {
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/pages/Authenticate.tsx]: done

File: template/src/components/chat-ui/index.tsx [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+  pendingPrivateNotification: any,
+  pendingPublicNotification: any,
+  lastCheckedPrivateState: any,
+  privateMessageCountMap: any,
+  setLastCheckedPublicState: React.Dispatch<SetStateAction<any>>,
+  setPrivateMessageLastSeen: React.Dispatch<SetStateAction<any>>,
+  setPrivateChatDisplayed: React.Dispatch<SetStateAction<any>>
+}
+
+const ChatUIContext = createContext((null as unknown) as ChatUIInterface);
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/components/chat-ui/index.tsx]: done

File: template/src/components/common/Error.tsx [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+type ErrorType = {
+  name: string;
+  message: string
+} 
+type ErrorContextType = {
+  error: ErrorType | undefined,
+  setGlobalErrorMessage: (error: any) => void,
+  resetError: () => void
+}
+const ErrorContext = React.createContext((null as unknown) as ErrorContextType);
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/components/common/Error.tsx]: done

File: template/src/components/dimension/DimensionContext.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+ information visit https://appbuilder.agora.io. 
+*********************************************
+*/
+import {createContext} from 'react';
+
+interface DimensionContextInterface {
+  getDimensionData: (width?: number, height?: number) => {dim:(number | boolean)[],isDesktop: boolean}
+}
+
+const DimensionContext = createContext((null as unknown) as DimensionContextInterface);
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/components/dimension/DimensionContext.ts]: done

File: template/src/components/precall/LocalMute.tsx [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+import { getFpeSubCmpConfig } from 'fpe-api';
+import { LocalUserContext } from '../../../agora-rn-uikit';
+
+const PreCallLocalMute: React.FC = () => {
+  const { PreCallLocalAudioMute, PreCallLocalVideoMute } = getFpeSubCmpConfig()
+  return (
+    <View style={style.precallControls}>
+      <LocalUserContext>
+        <View style={{ alignSelf: 'center' }}>
+          {PreCallLocalVideoMute && <PreCallLocalVideoMute />}
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/components/precall/LocalMute.tsx]: done

File: template/src/components/precall/PreCall.tsx [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+ information visit https://appbuilder.agora.io.
+*********************************************
+*/
+import React, { useContext, useEffect, useState} from 'react';
+import {View, StyleSheet, Platform} from 'react-native';
+import { getFpeSubCmpConfig } from 'fpe-api';
+import PreCallContext from './PreCallContext';
+import DimensionContext from '../dimension/DimensionContext';
+
+const Precall = (props: any) => {
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/components/precall/PreCall.tsx]: done

File: template/src/components/precall/PreCall.tsx [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    }
+  })
+  //TODO: move context to separete folder
+  //isWeb Platform.OS === 'web' move to utils
+  return (
+    <View style={style.main} onLayout={onLayout}>
+      {PreCallLogo && <PreCallLogo/>}
+      <View style={style.content}>
+        <View style={style.leftContent}>
+          {PreCallVideoPreview && <PreCallVideoPreview />}
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/components/precall/PreCall.tsx]: done

File: template/src/components/precall/PreCallContext.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+interface PreCallContextInterface {
+  username: string;
+  setUsername: React.Dispatch<React.SetStateAction<string>>;
+  setCallActive: React.Dispatch<React.SetStateAction<boolean>>;
+  queryComplete: boolean;
+  title: string;
+}
+
+const PreCallContext = createContext((null as unknown) as PreCallContextInterface);
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/components/precall/PreCallContext.ts]: done

File: template/src/components/precall/PreCallProvider.tsx [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+ accessed, used, modified, or distributed for any purpose without a license from Agora Lab, Inc.  
+ Use without a license or in violation of any license terms and conditions (including use for 
+ any purpose competitive to Agora Lab, Inc.’s business) is strictly prohibited. For more 
+ information visit https://appbuilder.agora.io. 
+*********************************************
+*/
+import React from 'react';
+import PreCallContext from './PreCallContext';
+
+const PreCallProvider = (props:any) => {
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/components/precall/PreCallProvider.tsx]: done

File: template/src/components/precall/index.tsx [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+import React from 'react';
+import PreCall from './PreCall';
+import PreCallLocalMute from './LocalMute';
+import PreCallVideoPreview from './VideoPreview';
+import PreCallSetName from './setName';
+import PreCallSelectDevice from './selectDevice';
+import {LocalAudioMute,LocalVideoMute} from '../../../agora-rn-uikit';
+import PreCallJoinBtn from './joinCallBtn'
+import PreCallTextInput from './textInput';
+const PreCallLocalAudioMute = () => <LocalAudioMute />;
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/components/precall/index.tsx]: done

File: template/src/components/precall/joinCallBtn.tsx [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+import React, { useContext } from 'react';
+import PrimaryButton from '../../atoms/PrimaryButton';
+import PreCallContext from './PreCallContext';
+
+const joinCallBtn: React.FC = () => {
+  const { username, queryComplete, setCallActive } = useContext(PreCallContext)
+  return (
+    <PrimaryButton
+        onPress={() => setCallActive(true)}
+        disabled={!queryComplete || username.trim() === ''}
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/components/precall/joinCallBtn.tsx]: done

File: template/src/components/precall/setName.tsx [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+
+import React from 'react';
+import { View, StyleSheet } from 'react-native'
+import { getFpeSubCmpConfig } from 'fpe-api';
+
+const setName: React.FC = () => {
+  const {PreCallJoinBtn, PreCallTextInput} = getFpeSubCmpConfig()
+  return (
+    <View style={[style.margin5Btm, { alignItems: 'center' }]}>
+      {PreCallTextInput && <PreCallTextInput/>}
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/components/precall/setName.tsx]: done

File: template/src/components/Controls.tsx [hunk 16]
```diff
@@ -25,13 +24,10 @@ import {
...
 import ScreenshareButton from '../subComponents/ScreenshareButton';
-// import ColorContext from './ColorContext';
-// import ChatContext from '../components/ChatContext';
-// import {SidePanelType} from '../subComponents/SidePanelEnum';
 import {controlsHolder} from '../../theme.json';
-import mobileAndTabletCheck from '../utils/mobileWebTest';
+import {isMobileWeb} from '../utils/common';
+import { VideoCallContext } from '../pages/VideoCall';
 
 const Controls = (props: any) => {
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/components/Controls.tsx]: done

File: template/src/utils/common.tsx [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+
+const isWeb = Platform.OS === 'web'
+const isAndroid = Platform.OS === 'android'
+const isIOS = Platform.OS === 'ios'
+const isMac = Platform.OS === 'macos'
+const isWindows = Platform.OS === 'windows'
+const isNative = isAndroid || isIOS
+const isDesktop = isMac || isWindows
+
+const isMobileWeb = () => {
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/utils/common.tsx]: done

File: template/src/pages/Create/ShareLink.tsx [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+
+type ShareLinkContextType = {
+  urlView: string;
+  urlHost: string;
+  pstn: string;
+  hostControlCheckbox: boolean;
+  joinPhrase: string
+  roomTitle: string
+}
+const ShareLinkContext = React.createContext((null as unknown) as ShareLinkContextType);
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/pages/Create/ShareLink.tsx]: done

File: template/src/subComponents/TextWithTooltip.tsx [hunk 19]
```diff
@@ -41,4 +41,4 @@ const style = {
  * Web and Desktop : using the TextWithToolTip - which have the browser tooltip 
  * Mobile and Mobile Web : using the TextWithToolTipNative - which have the custom tooltip using modal
  */
-export default mobileAndTabletCheck() ? TextWithToolTipNative : TextWithToolTip;
+export default isMobileWeb() ? TextWithToolTipNative : TextWithToolTip;
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/subComponents/TextWithTooltip.tsx]: done

File: template/src/components/Controls.tsx [hunk 20]
```diff
@@ -71,57 +57,24 @@ const Controls = (props: any) => {
...
-            style={{
-              textAlign: 'center',
-              marginTop: 5,
-              color: $config.PRIMARY_COLOR,
-            }}>
-            Video
-          </Text> */}
         </View>
-        {mobileAndTabletCheck() ? (
+        {isMobileWeb() ? (
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/components/Controls.tsx]: No difference. i think its now rename as mobileOrTablet for betting understanding

File: template/src/components/dimension/DimensionProvider.tsx [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+ information visit https://appbuilder.agora.io. 
+*********************************************
+*/
+import React from 'react';
+import {Dimensions} from 'react-native'
+import DimensionContext from './DimensionContext';
+
+const DimensionProvider = (props:any) => {
+
+  const getDimensionData = (width?: number, height?: number):{dim:(number | boolean)[],isDesktop: boolean} => {
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/src/components/dimension/DimensionProvider.tsx:18]: Typescript throwing error if i not specified the return type. that why i added it.

File: template/fpe-api/index.ts [hunk 22]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+let components: ComponentsInterface = {
+  PreCall: PreCall,
+  createMeetingScreen: Create,
+  joinMeetingScreen: Join,
+  videoCallScreen: VideoCall
+}
+
+let FpeApiConfig: FpeApiInterface = {
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/fpe-api/index.ts]: Now we are not mutating fpe object

File: template/fpe-api/index.ts [hunk 23]
```diff
@@ -0,0 +1,198 @@
...
+  }
+  FpeApiConfig.subcomponents = temp
+}
+
+const installCustomRoutes = (custom_routes: CustomRoutesInterface[] = []) => {
+  FpeApiConfig.custom_routes = custom_routes?.concat(getDefaultRoutes())
+}
+
+const installFPE = (config: FpeApiInterface) => {
+  installComponents(config?.components)
```
[the remark this answers is not in the document]
Comment (HariharanIT) [reply template/fpe-api/index.ts]: with current useFpe design. we are not needed pick method for now. if any need i will use this

**Pull request merged**